lib/elixir/lib/keyword.ex

```diff
@@ -449,9 +449,20 @@ defmodule Keyword do
   """
   @spec delete(t, key, value) :: t
   def delete(keywords, key, value) when is_list(keywords) and is_atom(key) do
-    :lists.filter(fn {k, v} -> k != key or v != value end, keywords)
+    do_delete(keywords, key, value, _deleted? = false)
+  catch
+    :not_deleted -> keywords
   end
 
+  defp do_delete([{key, value} | rest], key, value, _deleted?),
+    do: do_delete(rest, key, value, true)
+  defp do_delete([{_, _} = pair | rest], key, value, deleted?),
+    do: [pair | do_delete(rest, key, value, deleted?)]
+  defp do_delete([], _key, _value, _deleted? = true),
+    do: []
+  defp do_delete([], _key, _value, _deleted? = false),
+    do: throw(:not_deleted)
```

Comment:
Is this faster than returning `nil` here and doing `delete(keywords, key, value, false) || keywords` in the original function?

Reply:
We can't return `nil` since we have a body-recursive function. This would only set the final tail to `nil`

Reply:
Oh wow yes of course, sorry!

Reply:
But we can totally do that in the `pop` case, I'll see how it compares